### ISSUE TITLE
Add example template to reproduce lean (~100KB) binary size on clang/gcc

### ISSUE THREAD

**Issue by dellis23** [iree-org/iree#11516]
### Request description

The [IREE paper](https://arxiv.org/abs/2205.14479) describes a built runtime binary of ~100KB.  Using [this template](https://github.com/benvanik/iree-template-runtime-cmake/), we're able to get close -- around 250KB.  It would be nice to have clear instructions or a template for how a user can reproduce a binary as small as the one in the original paper.

### What component(s) does this issue relate to?

Runtime

### Additional context

Relevant discussion in discord starts [here](https://discord.com/channels/689900678990135345/892053976068001893/1049778076457906197).

**Comment by benvanik** [comment]
Note that at that scale the deployment type decides the binary size and if you want to get smaller you need to start working harder and it's very situational as to whether it's worth it.

Examples of things you can do to cut down on binary size (each can reduce binary size by ~10-50KB):
* use emit-c to statically link the host code into your binary instead of the dynamic VMFB loader and bytecode interpreter
* use static library output for device code to avoid the dynamic ELF loader on CPU
* use the inline HAL when running synchronously on single-core CPU devices to avoid the full pluggable HAL layer
* don't use helpers - avoid the generic iree_runtime_* APIs and instead the raw VM/HAL APIs (each helper is doing what you could do but with care for cases you may not have - when targeting tiny sizes it's almost always worth it to write 10 lines of code to do precisely what you need than 1 that may pull in additional deps to do things you don't)
* don't measure non-IREE things like IO/libc/etc and don't use optional IREE tooling helpers (like printing/parsing buffer views to stdout, etc) - string formatting code pulls in a lot of stuff and shouldn't be used in real apps (especially not size-sensitive ones)

The existing samples/static_library/ shows how to use the emit-c + static library device code and gets things pretty small but it's still using some helpers and the full HAL. It does probably need to be rewritten, though, as it's evolved somewhat randomly. It should still be useful for seeing how to do emit-c + static device libraries.

The TinyIREE paper made quite a few of these decisions because it makes sense in that specific deployment of statically linked, single-threaded, bare-metal, and CPU-only execution. You won't be able to hit the same size if your requirements don't match those, though nowadays it should be possible to get much smaller than the TinyIREE paper if your requirements do happen to match (the paper predates the inline HAL for example and that can easily save 40-50KB!).

**Comment by dellis23** [comment]
Makes sense, thanks.  Perhaps the options could be prioritized in terms of bang-for-buck and presented to the user in some way that's easy to understand and compare (ideally behind a flag, but it seems like a few of these would require rewriting the application itself).

**Comment by benvanik** [comment]
Yeah we definitely need the compiler flags for the size-optimized builds and some things like the inline HAL have no examples today but could easily get some. I think we can show off all the individual things in the main repo but getting combinations like in the TinyIREE paper are probably best as standalone things we can fully document/test/etc with hardware- and platform-specific frameworks (platform.io/renode/etc).
